packages/execution-environments/src/common/keyring.ts

```diff
@@ -21,29 +27,33 @@ export function wrapKeyring(
     throw new Error('Keyring not exported');
   }
 
-  const keyringHandler = ({
-    request,
-  }: {
-    request: JsonRpcRequest<unknown[]>;
-  }) => {
+  const keyringHandler = ({ request }: { request: JsonRpcRequest<Json[]> }) => {
     const { method, params } = request;
     if (!(method in keyring)) {
       throw new Error(`Keyring does not expose ${method}`);
     }
-    let args = params ?? [];
-    // @ts-expect-error TODO: Figure out how to type this better
-    const func = keyring[method].bind(keyring);
+    let args = (params ?? []) as unknown[];
+    const keyringMethod = keyring[method as keyof SnapKeyring];
+    assert(keyringMethod !== undefined);
+    const func = keyringMethod.bind(keyring);
     // Special case for registering events
     if (method === 'on') {
-      const data = args[0];
-      const listener = (listenerArgs: unknown) =>
-        notify({
+      const data = args[0] as Json;
+      const listener = (...listenerArgs: unknown[]) => {
+        assert(
+          isValidJson(listenerArgs),
+          new TypeError(
+            'Keyrings .on listener received non Json serializable value',
+          ),
+        );
+        return notify({
           method: 'SnapKeyringEvent',
           params: { data, args: listenerArgs },
         });
+      };
       args = [data, listener];
     }
-    return func(...args);
+    return (func as (..._: unknown[]) => unknown)(...args);
```

Comment:
What's going on here?

Reply:
`func` has a concrete type of `SnapKeyring[string]`, which a list of functions with conretized arguments.

Typescript disallows passing `...rest` arguments of unknown size into functions that don't take variable number of arguments. We need to cast `func` into a function that takes variable number of arguments.